File: src/de/handelsreg.ts
/**
 * Handelsregisternummer (German Company Register Number).
 *
 * German commercial register number assigned by the
 * local court (Amtsgericht). Format: register type
 * (HRA, HRB) + number, optionally with court name.
 *
 * HRA = Handelsregister Abteilung A (partnerships,
 * sole proprietors). HRB = Abteilung B (corporations).
 * GnR = Genossenschaftsregister.
 * PR = Partnerschaftsregister.
 * VR = Vereinsregister.
 *
 * No check digit; validation is structural only.
 *
 * @see https://de.wikipedia.org/wiki/Handelsregister_(Deutschland)
 */

import { clean } from "#util/clean";
import { err } from "#util/result";

import type { ValidateResult, Validator } from "../types";

const REGISTER_TYPES = new Set([
  "HRA",
  "HRB",
  "GNR",
  "PR",
  "VR",
]);

const compact = (value: string): string =>
  clean(value, " .").trim().toUpperCase();

/**
 * Parse out the register type and number from
 * various formats like "HRB 12345", "HRB12345",
 * "Amtsgericht München HRB 12345".
 */
const parse = (
  value: string,
): { type: string; number: string } | null => {
  const v = compact(value);
  // Match register type followed by digits
  const match = v.match(/(HRA|HRB|GNR|PR|VR)\s*(\d+)/);
  if (!match) return null;
  return { type: match[1]!, number: match[2]! };
};

const validate = (value: string): ValidateResult => {
  const parsed = parse(value);

  if (!parsed) {
    return err(
      "INVALID_FORMAT",
      "Handelsregisternummer must contain a register "
        + "type (HRA, HRB, GnR, PR, VR) followed "
        + "by a number",
    );
  }

  if (!REGISTER_TYPES.has(parsed.type)) {
    return err(
      "INVALID_COMPONENT",
      "Invalid register type",
    );
  }

  if (
    parsed.number.length < 1
    || parsed.number.length > 7
  ) {
    return err(
      "INVALID_LENGTH",
      "Register number must be 1 to 7 digits",
    );
  }

  return {
    valid: true,
    compact: `${parsed.type} ${parsed.number}`,
  };
};

const format = (value: string): string => {
  const parsed = parse(value);
  if (!parsed) return compact(value);
  return `${parsed.type} ${parsed.number}`;
};

/** German Company Register Number. */
const handelsreg: Validator = {
  name: "German Company Register Number",
  localName: "Handelsregisternummer",
  abbreviation: "HReg",
  aliases: [
    "Handelsregisternummer",
    "Handelsregister",
    "HRB",
    "HRA",
  ] as const,
  candidatePattern:
    "(?:HRA|HRB|GnR|PR|VR)\\s*\\d{1,7}",
  country: "DE",
  entityType: "company",
  sourceUrl:
    "https://de.wikipedia.org/wiki/Handelsregister_(Deutschland)",
  examples: ["HRB 12345"] as const,
  compact: (v: string) => {
    const parsed = parse(v);
    if (!parsed) return clean(v, " .").trim();
    return `${parsed.type} ${parsed.number}`;
  },
  format,
  validate,
};

export default handelsreg;
export { compact, format, validate };

File: crates/stdnum-core/src/validators/de/handelsreg.rs
//! German Company Register Number.

use crate::{
  ValidationResult, Validator, compact_without,
  types::{CountryCode, EntityType, ValidatorScope, ValidatorSpec},
  validators::common::{
    generate_from_examples, invalid_component, invalid_format, is_digits,
  },
};

const EXAMPLES: &[&str] = &["HRB 12345"];
const KINDS: &[&str] = &["HRA", "HRB", "GNR", "PR", "VR"];

pub static VALIDATOR: Validator = Validator::new(ValidatorSpec {
  id: "de.handelsreg",
  name: "German Company Register Number",
  local_name: "Handelsregisternummer",
  abbreviation: "HReg",
  aliases: &["Handelsregisternummer", "Handelsregister", "HRB", "HRA"],
  candidate_pattern: r"(?:HRA|HRB|GnR|PR|VR)\s*\d{1,7}",
  scope: ValidatorScope::Country(CountryCode::De),
  entity_type: EntityType::Company,
  source_url: Some(
    "https://de.wikipedia.org/wiki/Handelsregister_(Deutschland)",
  ),
  lengths: &[],
  examples: EXAMPLES,
  compact,
  format,
  validate,
  generate: Some(generate),
  parse: None,
});

#[must_use]
pub fn compact(value: &str) -> String {
  let compact = compact_without(value.trim(), &[' ', '-', '.']).to_uppercase();
  for kind in KINDS {
    if let Some(number) = compact.strip_prefix(kind)
      && (1..=7).contains(&number.len())
      && is_digits(number)
    {
      return format!("{kind} {number}");
    }
  }
  compact
}

#[must_use]
pub fn format(value: &str) -> String {
  compact(value)
}

pub fn validate(value: &str) -> ValidationResult {
  let value = compact(value);
  let Some((kind, number)) = value.split_once(' ') else {
    return Err(invalid_format());
  };
  if !KINDS.contains(&kind) {
    return Err(invalid_component());
  }
  if number.is_empty() || number.len() > 7 || !is_digits(number) {
    return Err(invalid_format());
  }
  Ok(value)
}

#[must_use]
pub fn generate() -> String {
  generate_from_examples(EXAMPLES, compact, validate)
}
